packages/fetchai/connections/prometheus/connection.py

```diff
@@ -0,0 +1,341 @@
+        elif message.performative == PrometheusMessage.Performative.UPDATE_METRIC:
+            response = await self._handle_update_metric(message)
+        else:  # pragma: nocover
+            self.logger.warning("Unrecognized performative for PrometheusMessage")
+            return
+
+        response_code, response_msg = cast(Tuple[int, str], response)
+
+        msg = dialogue.reply(
+            performative=PrometheusMessage.Performative.RESPONSE,
+            target_message=message,
+            code=response_code,
+            message=response_msg,
+        )
+        context = cast(EnvelopeContext, envelope.context)
+        envelope = Envelope(
+            to=msg.to,
+            sender=msg.sender,
+            protocol_id=msg.protocol_id,
+            message=msg,
+            context=context,
+        )
+        await self._send(envelope)
+
+    async def _handle_add_metric(self, message: PrometheusMessage) -> Tuple[int, str]:
+        """Handle add metric message.
+
+        :param message: the message to handle.
+        :return: the response code and response message.
+        """
+        if message.title in self.metrics:
+            response_code = 409
+            response_msg = "Metric already exists."
+        else:
+            metric_type = getattr(aioprometheus, message.type, None)
+            if metric_type is None:
+                response_code = 404
+                response_msg = f"{message.type} is not a recognized prometheus metric."
+            else:
+                self.metrics[message.title] = metric_type(
+                    message.title, message.description, message.labels
+                )
+                self._service.register(self.metrics[message.title])
+                response_code = 200
+                response_msg = (
+                    f"New {message.type} successfully added: {message.title}."
+                )
+
+        return response_code, response_msg
+
+    async def _handle_update_metric(
+        self, message: PrometheusMessage
+    ) -> Tuple[int, str]:
+        """Handle update metric message.
+
+        :param message: the message to handle.
+        :return: the response code and response message.
+        """
+        metric = message.title
+        if metric not in self.metrics:
+            response_code = 404
+            response_msg = f"Metric {metric} not found."
+        else:
+            update_func = getattr(self.metrics[metric], message.callable, None)
+            if update_func is None:
+                response_code = 400
+                response_msg = (
+                    f"Update function {message.callable} not found for metric {metric}."
+                )
+            else:
+                # Update the metric ("inc" and "dec" do not take "value" argument)
+                if message.callable in {"inc", "dec"}:
+                    update_func(message.labels)
+                else:
+                    update_func(message.labels, message.value)
```

Comment:
can we be sure this is a method? can it be a coroutine too?

Reply:
What if it fails?

Reply:
There is now a check to make sure that is a callable and implemented function. There may be some other obscure failure cases (e.g. invalid strings in names or labels), but I think we cover the most important ones now.

Reply:
Similar to above, this is the correct usage (not as a coroutine) based on my understanding.

Reply:
it's just a functions, not coroutines